benchmark/projection/include/create_data.hpp

```diff
@@ -0,0 +1,168 @@
+      npar[cell] = pdist(random);
+    }
+  }
+  return npar;
+}
+
+template <int nmode, typename T, typename Shape>
+static inline auto create_data(sycl::queue &Q, int ncell, int min_per_cell,
+                               int max_per_cell, int active_cells = -1) {
+  // Shove all the pointers in a vector so we can free them later
+
+  std::vector<void *> all_pointers;
+
+  // Par per cell
+  int *par_per_cell = sycl::malloc_device<int>(ncell, Q);
+  assert(par_per_cell);
+  all_pointers.push_back((void *)par_per_cell);
+  int max_row = 0;
+  std::vector<int> host_par_per_cell;
+  if (active_cells < 0 || active_cells == ncell) {
+    host_par_per_cell = dense_dist(ncell, min_per_cell, max_per_cell);
+    max_row =
+        *std::max_element(host_par_per_cell.begin(), host_par_per_cell.end());
+    Q.copy<int>(host_par_per_cell.data(), par_per_cell, ncell).wait();
+  } else {
+    host_par_per_cell =
+        simplesol_dist(ncell, active_cells, min_per_cell, max_per_cell);
+    max_row =
+        *std::max_element(host_par_per_cell.begin(), host_par_per_cell.end());
+    Q.copy<int>(host_par_per_cell.data(), par_per_cell, ncell).wait();
+  }
+  // DOFS
+  T *dofs =
+      sycl::malloc_device<T>(ncell * Shape::template get_ndof<nmode>(), Q);
+  assert(dofs);
+  all_pointers.push_back((void *)dofs);
+  Q.fill(dofs, T{0.0}, ncell * Shape::template get_ndof<nmode>())
+      .wait_and_throw();
+
+  // DOF offsets ndof per cell but scanned
+  auto host_offsets =
+      std::vector<int>(ncell, Shape::template get_ndof<nmode>());
+  std::exclusive_scan(host_offsets.begin(), host_offsets.end(),
+                      host_offsets.begin(), 0);
+  int *dof_offsets = sycl::malloc_device<int>(ncell, Q);
+  assert(dof_offsets);
+  all_pointers.push_back((void *)dof_offsets);
+  Q.copy<int>(host_offsets.data(), dof_offsets, ncell).wait_and_throw();
+
+  // Cell ids
+  int *cell_ids = sycl::malloc_device<int>(ncell, Q);
+  assert(cell_ids);
+  all_pointers.push_back((void *)cell_ids);
+  Q.parallel_for<>(ncell, [=](sycl::id<1> id) { cell_ids[id] = id; }).wait();
+  // Particle postion pointers
+  auto host_data_ptrs = std::vector<T **>(ncell, nullptr);
+  for (int i = 0; i < ncell; ++i) {
+    // allocate data arrays for positions
+    T *par_data[3] = {nullptr};
+    par_data[0] = sycl::malloc_device<T>(host_par_per_cell[i], Q);
+    assert(par_data[0]);
+    random_fill_array(par_data[0], host_par_per_cell[i], Q);
+    all_pointers.push_back((void *)par_data[0]);
+    if (Shape::dim > 1) {
+      par_data[1] = sycl::malloc_device<T>(host_par_per_cell[i], Q);
+      assert(par_data[1]);
+      random_fill_array(par_data[1], host_par_per_cell[i], Q);
+      all_pointers.push_back((void *)par_data[1]);
+    }
+    if (Shape::dim > 2) {
+      par_data[2] = sycl::malloc_device<T>(host_par_per_cell[i], Q);
+      random_fill_array(par_data[2], host_par_per_cell[i], Q);
+      assert(par_data[2]);
+      all_pointers.push_back((void *)par_data[2]);
+    }
```

Comment:
This looks like a loop over dimensions done with if statements.

Reply:
Ha ha -  I'm an idiot. I'll fix this and we can never mention it again.